### PR TITLE
Refactored apply/detach of ControlTemplates.

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

This PR builds on #8600 to improve performance when switching `Theme`s on a control. Previously we were simply removing _all_  styling from the control (whether related to the `ControlTheme` or not) and then re-applying all styles and control themes on the next layout pass.

This PR allows us to just remove the value frames that come from the control theme being changed, leaving all of the other applied styles as they are. To do this we add a `ValueFrame.FramePriority` property which encodes both the `BindingPriority` and source of the frame (style, control theme, templated parent control theme) on each value frame. Then, when the `Theme` changes we can simply remove these frames and queue the control theme to be re-applied on the next layout pass.

## Breaking Changes

Also made a few API changes:

- Renames `BindingPriority.TemplatedParent` to `Template` as it no longer represents just `TemplatedParent` bindings, but every value or binding created by a template (a `TemplatedParent` entry is left in there for backwards compat so shouldn't be breaking)
- Removed `IStyleable.DetachStyles` and protected `StyledElement.InvalidateStyles` as I wasn't happy with making detaching styles available in a public API (it should only be needed for internal unit tests)

## Benchmarks

I added a benchmark which switches control theme in the presence of a bunch of styles defined in a nested hierarchy of controls:

before

|              Method |     Mean |   Error |  StdDev |   Gen 0 |  Gen 1 | Allocated |
|-------------------- |---------:|--------:|--------:|--------:|-------:|----------:|
| Change_ControlTheme | 354.4 us | 3.50 us | 4.17 us | 11.7188 | 1.9531 |     49 KB |

after

|              Method |     Mean |   Error |  StdDev |  Gen 0 |  Gen 1 | Allocated |
|-------------------- |---------:|--------:|--------:|-------:|-------:|----------:|
| Change_ControlTheme | 226.6 us | 1.22 us | 1.08 us | 9.2773 | 2.1973 |     39 KB |

So seeing a 40% speedup there,

Depends on #9411

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0026436-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0026474-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: src/Avalonia.Base/StyledElement.cs [hunk 1]
```diff
@@ -886,70 +895,81 @@ private void ClearLogicalParent(IList children)
...
-                }
+            if (propagate)
+            {
+                e ??= ResourcesChangedEventArgs.Empty;
+                NotifyChildResourcesChanged(e);
             }
         }
 
-        private void DetachStylesFromThisAndDescendents(IReadOnlyList<StyleBase> styles)
+        private static IReadOnlyList<Style>? FlattenStyles(IReadOnlyList<IStyle> styles)
```
Comment (maxkatz6) [inline src/Avalonia.Base/StyledElement.cs:933]: I am pretty sure we can return a pooled list from here and dispose after "RemoveFrames" was called.

**Review by maxkatz6 — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0026513-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Pull request merged**